lib/i18n/backend/lazy_load.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+module I18n
+  module Backend
+    # Backend that lazy loads translations based on the current locale. This
+    # implementation avoids loading all translations up front. Instead, it only
+    # loads the translations that belong to the current locale. This offers a
+    # performance incentive in local development and test environments for
+    # applications with many translations for many different locales. It's
+    # particularly useful when the application only refers to a single locales'
+    # translations at a time (ex. A Rails workload).  The implementation
+    # identifies which translation files from the load path belong to the
+    # current locale by pattern matching against their path name.
+    #
+    # Specifically, a translation file is considered to belong to a locale if:
+    # a) the filename is in the I18n load path
+    # b) the filename ends in a supported extension (ie. .yml, .json, .po, .rb)
+    # c1) the filename starts with the locale identifier
+    #    OR
+    # c2) the path contains a component equal to the locale identifier
+    #
+    # Examples:
+    # Valid files that will be selected by this backend:
+    #
+    # "files/locales/en-translation.yml" (Selected for locale "en")
+    # "files/locales/fr/translation.po"  (Selected for locale "fr")
+    #
+    # Invalid files that won't be selected by this backend:
+    #
+    # "files/locales/translation-file"
+    # "files/locales/en-translation.unsupported"
+    # "files/locales/french/translation.yml"
+    #
+    # The implementation uses this assumption to defer the loading of
+    # translation files until the current locale actually requires them.
+    #
+    # Note: This backend isn't designed for production environments or other
+    # environments where eager loading all translations is recommended.
+    #
+    # To use the LazyLoad backend instantiate it and set it to the I18n module.
+    # You can configure lazy loaded backends through the initializer or backends
+    # accessor:
+    #
+    #   # configure I18n backend to use LazyLoad Backend
+    #   I18n.backend = I18n::Backend::LazyLoad.new
+    #
+    class LazyLoad < Simple
+      # Returns whether the current locale is initialized.
+      def initialized?
+        initialized_locales[I18n.locale]
+      end
+
+      # Clean up translations and uninitialize all locales.
+      def reload!
+        @initialized_locales = nil
+        @translations = nil
+      end
+
+      def eager_load!
+        raise UnsupportedMethod.new(__method__, self.class)
+      end
```

Comment:
So second though on this in the context of our CI (and most people CI now that Rails has eagerload by default on CI).

It might make sense to eager_load the "default locale" (typically `:en`) as it's what the vast majority of tests will end up using.

It's however tricky to reconcile the safety of preventing this from being used in production, with this extra feature. So I'm not suggesting to act on this, just braindumping a concern for which I have no solution yet.

Reply:
Actually, on CI it might make sense to eager load all locales regardless of wether they are used as it at the very least test that they don't contain syntax errors and such. So ignore the above.

Reply:
I wonder if we should encourage folks to use the same backend for both production and CI, like we do in Core? That way we ensure CI loads everything the same way production would, and if they have their own extensions / are using some of the pluggable I18n features those will be tested in CI.

I'm also wondering how we'll expect users to configure this. Should it just be a setting on the gem that can be opted into (ie. `I18n.lazy_load_in_test = true`), or are we expecting users to prepend the `LazyLoad` module in an initializer in their application?

Reply:
Yeah, in the context of the main question here, which is how to map a file name to a locale, and how to answer it is correct, I think it would make sense for this backend to actually have two use mode.

One lazy like today, and one eager loading mode that behave a bit like the regular one, but also make sure the assumptions about the filename to locale mapping was correct.

It would need to be renamed though, because it it wouldn't always be lazy.